python/dgl/sampling/labor.py

```diff
@@ -0,0 +1,276 @@
+    >>> g_edges = g.all_edges(form='all', etype=('drug', 'interacts', 'drug'))
+    (tensor([0, 0, 1, 1, 3, 2]), tensor([1, 2, 0, 1, 2, 0]), tensor([0, 1, 2, 3, 4, 5]))
+    >>> excluded_edges  = {('drug', 'interacts', 'drug'): g_edges[2][:3]}
+    >>> sg = dgl.sampling.sample_labors(g, {'drug':[0, 1]}, 3, exclude_edges=excluded_edges)
+    >>> sg.all_edges(form='all', etype=('drug', 'interacts', 'drug'))
+    (tensor([2, 1]), tensor([0, 1]), tensor([0, 1]))
+    >>> sg.has_edges_between(g_edges[0][:3],g_edges[1][:3],etype=('drug', 'interacts', 'drug'))
+    tensor([False, False, False])
+
+    """
+    if F.device_type(g.device) == 'cpu' and not g.is_pinned():
+        frontier, importances = _sample_labors(
+            g, nodes, fanout, random_seed, edge_dir=edge_dir, prob=prob,
+            importance_sampling=importance_sampling, copy_ndata=copy_ndata,
+            copy_edata=copy_edata, exclude_edges=exclude_edges)
+    else:
+        frontier, importances = _sample_labors(
+            g, nodes, fanout, random_seed, edge_dir=edge_dir, prob=prob,
+            importance_sampling=importance_sampling, copy_ndata=copy_ndata, copy_edata=copy_edata)
+        if exclude_edges is not None:
+            eid_excluder = EidExcluder(exclude_edges)
+            if importance_sampling != 0:
+                frontier, importances = eid_excluder(frontier, importances)
+            else:
+                frontier = eid_excluder(frontier)
+    return (frontier, importances) if output_device is None else (frontier.to(output_device), list(map(lambda x: x.to(output_device), importances)))
+
+def _sample_labors(g, nodes, fanout, random_seed, edge_dir='in', prob=None,
+                      importance_sampling=0, copy_ndata=True, copy_edata=True, exclude_edges=None):
+    if not isinstance(nodes, dict):
+        if len(g.ntypes) > 1:
+            raise DGLError("Must specify node type when the graph is not homogeneous.")
+        nodes = {g.ntypes[0] : nodes}
+
+    nodes = utils.prepare_tensor_dict(g, nodes, 'nodes')
+    if len(nodes) == 0:
+        raise ValueError(
+            "Got an empty dictionary in the nodes argument. "
+            "Please pass in a dictionary with empty tensors as values instead.")
+    ctx = utils.to_dgl_context(F.context(next(iter(nodes.values()))))
+    nodes_all_types = []
+    NIDs_all_types = []
+    for ntype in g.ntypes:
+        if ntype in nodes:
+            nodes_all_types.append(F.to_dgl_nd(nodes[ntype]))
+            try:
+                if len(g.ntypes) > 1:
+                    NIDs_all_types.append(F.to_dgl_nd(g.ndata[NID][ntype]))
+                else:
+                    NIDs_all_types.append(F.to_dgl_nd(g.ndata[NID]))
+            except:
+                NIDs_all_types.append(nd.array([], ctx=ctx))
+        else:
+            nodes_all_types.append(nd.array([], ctx=ctx))
+            NIDs_all_types.append(nd.array([], ctx=ctx))
```

Comment:
This seems to be reading the "original node IDs" of the graph.  Why is that necessary?

Reply:
This is so that if we are sampling from a local graph, then global id's are used so that different machines can roll same random numbers for the same vertices. Same reason as using a random seed. Open to another suggestion to handle this.

Reply:
There can be an option to disable this so that there is less overhead in terms of reading from an extra array.

Reply:
> This is so that if we are sampling from a local graph, then global id's are used so that different machines can roll same random numbers for the same vertices.

I don't quite understand here.  By saying "different machines roll the same random numbers for the same vertices", you imply that the graph is replicated across all the machines?  If so, the "original node IDs" would not be necessary?

Reply:
The graph doesn't have to be replicated, it can be partitioned also. Cooperative minibatching PR partitions a graph to different GPUs and can run LABOR sampling on multiple GPUs with same outcome as if it was run on single machine with the same outcome.

Reply:
When one partitions the graph and also renames the vertices locally, then this functionality is needed to roll same random numbers for same vertices with different names in different partitions.

Reply:
So the necessity of NID is also a part of the cooperative minibatching?  My general opinion is that the logic of cooperative minibatching should go to the cooperative minibatching PR (or another PR about cooperative minibatching on LABOR sampling).

Reply:
Rolling a single random number for a vertex is crucial for labor to work for the scenarios vertices are renamed, for example when one is sampling from multiple subgraphs when the user passes relabel_nodes=True. If a batch were to consist of sampled subgraphs from different such subgraphs, then this NID logic is needed for labor sampling to work globally. Otherwise, its scope would be limited to each of those subgraphs, and it would sample more than it could have without this option.

Reply:
> If a batch were to consist of sampled subgraphs from different such subgraphs, then this NID logic is needed for labor sampling to work globally.

Is this a common use case for LABOR sampling (without considering cooperative minibatching)?

Reply:
I haven't figured out all the possible use cases yet. There may be any number of reasons a user might want to have such a sampling scheme. Since this is a new algorithm, I don't know how the users might want to take advantage of such a functionality but this logic is needed if we want the users to have this kind of flexibility.

Renaming the vertices of a subgraph from the original graph has the advantage that all the vertex ids become smaller. This might make the overall code faster also, for example radix sorting in cuda needs more passes to sort, the larger the vertex ids. Operations with the subgraph such as SpMM also become faster since accesses are more closely located in memory.

The addition of this feature harms the users of DGL in no way, but brings them flexibility in what additional things they can do. If you insist, I can add an option to use the NIDs, which could be turned off by default, and the users would have to manually opt-in if they want NIDs to be used.

Reply:
If we don't have an existing use case, we risk over-designing the API, and it complicates the implementation and maintenance of the code in the future (particularly by other people).  We (the DGL team) made the same mistakes before and it took a lot of effort to refactor out the unnecessary bits and further optimize the codebase.  So now we are trying to justify every design decision.

In this case, I think you can keep it as long as you justify the necessity with actual use cases (cooperative minibatching can be one for example).  It probably goes in the effort of having the informal post on the design decisions.

Reply:
For the NID argument, I think we can remove it from the Python API since the cooperative minibatching user-facing API is not settled.  We can still keep it in C++ interface since it's related to your future cooperative minibatching PR (with a comment describing why we need it).

Reply:
Disabled passing NIDs to the C++ side as per your suggestion along with a comment on what it was useful for.

Reply:
The default argument for the nids in C++ side is NullArray() and it is the last argument in the API, so the user doesn't even have to pass anything for the nids argument in the C++ api.

Reply:
I have done the same thing with the random seed, it is the second to last argument, and passing nothing for random_seed in the c++ API makes the code use a random seed that it gets from the DGL PRNG. Python side does the same thing, calling set_seed with no arguments sets the seed to a number it gets from the DGL PRNG.